File: src/Runtime/Sentence/SentenceWhile.cs
/*
	PeakScript: https://github.com/Jenocn/PeakScript
	By Jenocn: https://jenocn.github.io
*/

namespace peak.interpreter {
	class SentenceWhile : SentenceReturn {
		private SentenceExpression _expression = null;
		private Sentence _sentence = null;
		public SentenceWhile(SentenceExpression expression, Sentence sentence) {
			_expression = expression;
			_sentence = sentence;
		}
		public override ExecuteResult Execute(Space space) {
			if (_sentence) {
				var tempSpace = new Space(SpaceType.Loop, space);
				while (true) {
					if (!IsSuccess(_expression.Execute(space))) {
						ErrorLogger.LogRuntimeError(ErrorRuntimeCode.While, "The condition expression execute failed!");
						return ExecuteResult.Failed;
					}
					if (!ValueTool.ToLogic(_expression.value)) {
						break;
					}
					tempSpace.Clear();
					var ret = _sentence.Execute(tempSpace);
					if (!IsSuccess(ret)) {
						ErrorLogger.LogRuntimeError(ErrorRuntimeCode.While, "The sentence execute failed!");
						return ExecuteResult.Failed;
					}
					if (ret == ExecuteResult.Break) {
						break;
					}
					if (ret == ExecuteResult.Return) {
						SetReturnValue((_sentence as SentenceReturn).returnValue);
						tempSpace.Clear();
						return ExecuteResult.Return;
					}
				}
				tempSpace.Clear();
			} else {
				while (true) {
					if (!IsSuccess(_expression.Execute(space))) {
						ErrorLogger.LogRuntimeError(ErrorRuntimeCode.While, "The expression execute failed!");
						return ExecuteResult.Failed;
					}
					if (!ValueTool.ToLogic(_expression.value)) {
						break;
					}
				}
			}
			return ExecuteResult.Successed;
		}
	}
} // namespace peak.interpreter

File: src/runtime/sentence/sentence_while.cpp
#include "sentence_while.h"
#include "runtime/value/value_tool.h"
#include "sentence_expression.h"

using namespace peak;

SentenceWhile::SentenceWhile(std::shared_ptr<SentenceExpression> expression, std::shared_ptr<Sentence> sentence)
	: _expression(expression), _sentence(sentence) {
}
ExecuteResult SentenceWhile::Execute(std::shared_ptr<Space> space) {
	if (_sentence) {
		auto tempSpace = std::make_shared<Space>(SpaceType::Loop, space);
		while (true) {
			if (!IsSuccess(_expression->Execute(space))) {
				ErrorLogger::LogRuntimeError(ErrorRuntimeCode::While, "The condition expression execute failed!");
				return ExecuteResult::Failed;
			}
			if (!ValueTool::ToLogic(_expression->GetValue().get())) {
				break;
			}
			tempSpace->Clear();
			auto ret = _sentence->Execute(tempSpace);
			if (!IsSuccess(ret)) {
				ErrorLogger::LogRuntimeError(ErrorRuntimeCode::While, "The sentence execute failed!");
				return ExecuteResult::Failed;
			}
			if (ret == ExecuteResult::Break) {
				break;
			}
			if (ret == ExecuteResult::Return) {
				SetReturnValue(std::static_pointer_cast<SentenceReturn>(_sentence)->GetReturnValue());
				tempSpace->Clear();
				return ExecuteResult::Return;
			}
		}
		tempSpace->Clear();

	} else {
		while (true) {
			if (!IsSuccess(_expression->Execute(space))) {
				ErrorLogger::LogRuntimeError(ErrorRuntimeCode::While, "The expression execute failed!");
				return ExecuteResult::Failed;
			}
			if (!ValueTool::ToLogic(_expression->GetValue().get())) {
				break;
			}
		}
	}
	return ExecuteResult::Successed;
}